File: set_finder/poc/set_finder/shading_detector.py
import cv2
import numpy as np
from cards import Shading
from matplotlib import pyplot as plt

CANNY_TH1, CANNY_TH2 = 40, 120

GREEN_COLOR = (0, 255, 0)
THICKNESS = 2
ALL_CONTOURS_MODE = -1

class ShadingDetector:
    def __init__(self, cardPreprocessor):
        self._shading = Shading.UNKNOWN

        erodeImg = cardPreprocessor.getErodeImg()
        contours = cardPreprocessor.getContours()

        erodeImg = cv2.bilateralFilter(erodeImg, 7, 40, 40)

        grayImg = cv2.cvtColor(erodeImg, cv2.COLOR_BGR2GRAY)
        x, y, w, h = cv2.boundingRect(contours[0])
        roi = grayImg[y-1: y+1 + h, x-1 : x+1 + w]

        canny_roi = cv2.Canny(roi, CANNY_TH1, CANNY_TH2)
        roi_contours, _ = cv2.findContours(
            canny_roi, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE
        )

        # print(len(roi_contours))

        # cv2.drawContours(
        #     roi,
        #     roi_contours,
        #     ALL_CONTOURS_MODE,
        #     GREEN_COLOR,
        #     THICKNESS,
        # )

        # cv2.imshow("roi", roi)
        # while True:
        #     if cv2.waitKey(50) == 27:
        #         cv2.destroyAllWindows()
        #         break

        # plt.plot(hist, color="b")
        # plt.xlim([0, 64])
        # plt.show()

        # TODO: filter contours
        # for c in roi_contours:
        #     print(cv2.arcLength(c, closed=True))

        hist = cv2.calcHist([roi], [0], None, [64], [0, 256])

        reversedHist = hist[::-1]

        left_x = np.argmax(hist >= 100)
        left_x_2 = left_x + np.argmax(hist[left_x:] <= 100)

        right_x = np.argmax(reversedHist >= 100)
        right_x_2 = right_x + np.argmax(reversedHist[right_x:] <= 100)

        left_sum = sum(hist[left_x - 2 : left_x_2 + 2])
        right_sum = sum(reversedHist[right_x - 2 : right_x_2 + 2])

        contour_area = cv2.contourArea(contours[0])
        outer_pixels = (w+2)*(h+2) - contour_area

        if len(roi_contours) <= 2: # left_sum >= 0.75*abs(contour_area): # works with 0.6 - 0.9
            self._shading = Shading.SOLID
        elif (right_sum-outer_pixels) >= 0.40*(abs(contour_area)-left_sum): # works with 0.3 - 0.5
            self._shading = Shading.OUTLINED
        else:
            self._shading = Shading.STRIPED


    def getShading(self):
        return self._shading

File: set_finder/core/src/detector/shading_detector.cpp

#include "detector/shading_detector.h"

#include <opencv2/imgproc.hpp>

#include <algorithm>
#include <functional>
#include <numeric>

#include "utils/logger.h"

namespace
{
// Histogram parameters
constexpr int MAX_HIST_VALUE = 256;
const int hist_size[] = {64};
const float hranges[] = {0, static_cast<float>(MAX_HIST_VALUE)};
const float* ranges[] = {hranges};
const int* channels{0}; // NOLINT

constexpr int DIAMETER = 7;
constexpr double SIGMA_COLOR = 40;
constexpr double SIGMA_SPACE = 40;
constexpr int CANNY_TH1 = 40;
constexpr int CANNY_TH2 = 120;

constexpr uint16_t SUM_RANGE = 2;

constexpr double OUTLINED_TH = 0.4; // works with 0.3 - 0.5
constexpr uint16_t HIST_TH = 100;
} // namespace

Feature ShadingDetector::detect_feature(const CardPreprocessor& preprocessor) const
{
    const auto contours = preprocessor.get_contours();
    if(contours.empty())
    {
        return Shading::UNKNOWN;
    }

    const auto gray_card = generate_gray_card(preprocessor);

    const auto bounding_rect = cv::boundingRect(contours[0]);
    const auto extended_rect = cv::Rect(bounding_rect.x - 1,
                                        bounding_rect.y - 1,
                                        bounding_rect.width + 2,
                                        bounding_rect.height + 2);

    if(extended_rect.x < 0 || extended_rect.y < 0 ||
       (extended_rect.x + extended_rect.width) >= gray_card.size().width ||
       (extended_rect.y + extended_rect.height) >= gray_card.size().height)
    {
        return Shading::UNKNOWN;
    }

    const auto roi = gray_card(bounding_rect);
    const auto extended_roi = gray_card(extended_rect);

    Image canny_card{};
    cv::Canny(extended_roi, canny_card, CANNY_TH1, CANNY_TH2);

    // TODO: filter simillar arc-length wise contours
    Contours roi_contours{};
    cv::findContours(canny_card, roi_contours, {}, cv::RETR_TREE, cv::CHAIN_APPROX_SIMPLE);

    cv::Mat histogram{};
    cv::calcHist(&roi, 1, channels, {}, histogram, 1, hist_size, ranges);

    const auto [left_x, right_x] =
        get_hist_range(histogram, 0, hist_size[0], std::greater_equal<uint16_t>());
    const auto [left_x_2, right_x_2] =
        get_hist_range(histogram, left_x, right_x, std::less_equal<uint16_t>());

    const auto is_valid = [&](const auto& x) {
        return 0 < (x - SUM_RANGE) && hist_size[0] > (x + SUM_RANGE);
    };

    if(!is_valid(left_x_2) || !is_valid(right_x_2))
    {
        return Shading::UNKNOWN;
    }

    const auto left_sum = std::accumulate(
        histogram.begin<float>(), histogram.begin<float>() + left_x_2 + SUM_RANGE, 0);

    const auto right_sum = std::accumulate(
        histogram.begin<float>() + right_x_2 - SUM_RANGE, histogram.end<float>(), 0);

    const auto contour_area = std::abs(cv::contourArea(contours[0]));
    const auto outer_pixels = bounding_rect.width * bounding_rect.height - contour_area;

    if(roi_contours.size() <= 2) // (left_sum >= 0.75 * contour_area) // works with 0.6 - 0.9
    {
        return Shading::SOLID;
    }
    else if((right_sum - outer_pixels) >= OUTLINED_TH * (abs(contour_area) - left_sum))
    {
        return Shading::OUTLINED;
    }

    return Shading::STRIPED;
}

Image ShadingDetector::generate_gray_card(const CardPreprocessor& preprocessor) const
{
    const auto card = preprocessor.get_filtered_card();

    Image bilateral_card{};
    cv::bilateralFilter(card, bilateral_card, DIAMETER, SIGMA_COLOR, SIGMA_SPACE);

    Image gray_card{};
    cv::cvtColor(bilateral_card, gray_card, cv::COLOR_BGR2GRAY);
    return gray_card;
}

template <typename Comparator>
std::pair<uint16_t, uint16_t> ShadingDetector::get_hist_range(cv::Mat1f histogram,
                                                              uint16_t begin,
                                                              uint16_t end,
                                                              Comparator comparator) const
{
    uint16_t left_x{}, right_x{};
    for(uint16_t i = begin; i < end; ++i)
    {
        const auto histogram_value = static_cast<uint32_t>(*histogram[i]);
        if(comparator(histogram_value, HIST_TH))
        {
            if(left_x == 0)
            {
                left_x = i;
            }
            right_x = i;
        }
    }

    return {left_x, right_x + 1};
}
